File: python/classes/inheritance/multiple_inheritance.py
# Copyright (C) 2019-2026 Katarzyna Krasińska
# Languagium - https://github.com/katheroine/languagium
# Licensed under GPL-3.0 - see LICENSE.md
# SPDX-License-Identifier: GPL-3.0-or-later

class Donkey:
    def speak(self):
        print("Bray, bray!")

class Horse:
    def speak(self):
        print("Neigh, neigh!")

class Mule(Donkey, Horse):
    def speak(self):
        print("Bray, neigh!")

    def pretend_mother(self):
        Horse.speak(self)

    def pretend_father(self):
        Donkey.speak(self)

    def pretend_parent(self):
        super().speak()

mule_ernest = Mule()

mule_ernest.speak()
mule_ernest.pretend_mother()
mule_ernest.pretend_father()
mule_ernest.pretend_parent()

File: c++/classes/inheritance/multiple_inheritance.cpp
/*
 * Copyright (C) 2019-2026 Katarzyna Krasińska
 * Languagium - https://github.com/katheroine/languagium
 * Licensed under GPL-3.0 - see LICENSE.md
 */
// SPDX-License-Identifier: GPL-3.0-or-later

#include <iostream>

using std::string, std::cout, std::endl, std::boolalpha, std::noboolalpha;

class Donkey {
public:
  void speak() {
    cout << "Bray, bray!" << endl;
  }
};

class Horse {
public:
  void speak() {
    cout << "Neigh, neigh!" << endl;
  }
};

class Mule: Donkey, Horse {
public:
  void speak() {
    cout << "Bray, neigh!" << endl;
  }

  void pretend_mother() {
    Horse::speak();
  }

  void pretend_father() {
    Donkey::speak();
  }
};

int main() {
  Mule mule_ernest;

  mule_ernest.speak();
  mule_ernest.pretend_mother();
  mule_ernest.pretend_father();
}
